packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2057,6 +2086,34 @@ export class KeyringController extends BaseController<
     this.messagingSystem.publish(`${name}:unlock`);
   }
 
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and save the keyrings to state after it, or rollback to their
+   * previous state in case of error.
+   *
+   * @param fn - The function to execute.
+   * @returns The result of the function.
+   */
+  async #persistOrRollback<T>(fn: MutuallyExclusiveCallback<T>): Promise<T> {
+    return this.#withControllerLock(async ({ releaseLock }) => {
+      const currentSerializedKeyrings = await this.#getSerializedKeyrings();
```

Comment:
Thoughts on storing the current password state as well? I don't think this can be atomic unless we handle that too.

Whether or not the password is set was a significant factor in tracing this code to understand how the empty vault bug had occurred. We can have situations where the unlock fails partway through, but the password remains set so later persistence operations still succeed despite the vault being in a "locked" state apart from the password field.

Reply:
With the controller state, password, and keyrings accounted for, I think that's all of the mutable state. The rest of the instance variables are either static from the point of construction, or just used for diagnostic logging (`unsupportedKeyrings`).

Reply:
Oh definitely! added in [6c16b55](https://github.com/MetaMask/core/pull/4192/commits/6c16b55ea90797690b20fd27d0fe176a69964471)